arbor/hardware/node_info.cpp

```diff
@@ -1,19 +1,51 @@
-#include <algorithm>
+#include <thread>
+
+#ifdef ARB_HAVE_GPU
+#include <cuda_runtime.h>
+#endif
+
+// TODO: C++17 use __has_include(<unistd.h>)
+#if defined(__unix__) || defined(__APPLE__) && defined(__MACH__)
+#include <unistd.h>
+#endif
 
 #include "affinity.hpp"
-#include "gpu.hpp"
 #include "node_info.hpp"
 
 namespace arb {
 namespace hw {
 
-// Return a node_info that describes the hardware resources available on this node.
-// If unable to determine the number of available cores, assumes that there is one
-// core available.
-node_info get_node_info() {
-    auto res = num_cores();
-    unsigned ncpu = res? *res: 1u;
-    return {ncpu, num_gpus()};
+unsigned node_gpus() {
+#ifdef ARB_HAVE_GPU
+    int n;
+    if (!cudaGetDeviceCount(&n)) {
```

Comment:
Compare explicitly to `CUDA_SUCCESS`

Reply:
`CUDA_SUCCESS` (and success return values in pretty much every C or C++ library) is defined to be zero, and will never change. Similarly `MPI_SUCCESS`. I can't see how making the explicit comparison is helpful, and in fact makes it look like these values might possibly not be zero, which is misleading.

The convention that a zero return from a function (that's not returning a pointer) indicates success is really well established, and there is any amount of code along the lines of `if (some_library_call()) { deal_with_errror(); }`

Reply:
I don't think this is a good argument. You are assuming the value of `CUDA_SUCCESS`, which is not explicitly defined in the documentation anywhere. Why not state what is intended in the code?

Of course, it won't change, but it is a big undocumented leap from `cudaGetDeiceCount==CUDA_SUCCESS` to `!cudaGetDeiceCount`.